fix: Avoid writing on docs pipeline

igamigo opened the pull request:
Looks like the docs were failing to build for `miden-client` due to the same error we have seen before, related to the pipeline's filesystem being read-only: https://docs.rs/crate/miden-client/0.6.0/builds/1507411. The problem seems to come specifically from `miden-lib`, but I also added an early return for the prover service as well since that would also likely fail.

bobbinth reviewed (approved):
Looks good! Thank you!

PhilippGackstatter reviewed (changes requested):
This would not build the `kernel_v0.rs` and other files which are written directly into `src` but which are used in the crates, and building the docs locally in the docs.rs environment confirms it's failing, here's the output:

<details>
  <summary>Build errors (long output)</summary>
<pre><code>
# rustc version
rustc 1.84.0-nightly (90ab8eaed 2024-11-14)# docs.rs version
docsrs 0.6.0 (e2e7e30 2024-11-15)# build log
[INFO] running `Command { std: "docker" "create" "-v" "/home/philipp/git/docs.rs/.workspace/builds/miden-lib-0.6.1/target:/opt/rustwide/target:rw,Z" "-v" "/home/philipp/git/docs.rs/.workspace/builds/miden-lib-0.6.1/source:/opt/rustwide/workdir:ro,Z" "-v" "/home/philipp/git/docs.rs/.workspace/cargo-home:/opt/rustwide/cargo-home:ro,Z" "-v" "/home/philipp/git/docs.rs/.workspace/rustup-home:/opt/rustwide/rustup-home:ro,Z" "-e" "SOURCE_DIR=/opt/rustwide/workdir" "-e" "CARGO_TARGET_DIR=/opt/rustwide/target" "-e" "DOCS_RS=1" "-e" "CARGO_HOME=/opt/rustwide/cargo-home" "-e" "RUSTUP_HOME=/opt/rustwide/rustup-home" "-w" "/opt/rustwide/workdir" "-m" "3221225472" "--user" "1000:1000" "--network" "none" "ghcr.io/rust-lang/crates-build-env/linux-micro" "/opt/rustwide/cargo-home/bin/cargo" "+nightly" "rustdoc" "--lib" "-Zrustdoc-map" "--config" "build.rustdocflags=[\"--cfg\", \"docsrs\", \"-Z\", \"unstable-options\", \"--emit=invocation-specific\", \"--resource-suffix\", \"-20241114-1.84.0-nightly-90ab8eaed\", \"--static-root-path\", \"/-/rustdoc.static/\", \"--cap-lints\", \"warn\", \"--extern-html-root-takes-precedence\"]" "--offline" "-Zunstable-options" "--config=doc.extern-map.registries.crates-io=\"https://docs.rs/{pkg_name}/{version}/x86_64-unknown-linux-gnu\"" "-Zrustdoc-scrape-examples" "--target" "x86_64-unknown-linux-gnu", kill_on_drop: false }`
[INFO] [stdout] 9e2f96a3fb5156c28d1dd0a00b7c3d40424d15fe363836908b8b066285b45547
[INFO] running `Command { std: "docker" "start" "-a" "9e2f96a3fb5156c28d1dd0a00b7c3d40424d15fe363836908b8b066285b45547", kill_on_drop: false }`
[INFO] [stderr] warning: target filter specified, but no targets matched; this is a no-op
[INFO] [stderr]  Documenting miden-lib v0.6.1 (/opt/rustwide/workdir)
[INFO] [stderr] error: couldn't read `/opt/rustwide/target/x86_64-unknown-linux-gnu/debug/build/miden-lib-cfcaa074282a00e5/out/assets/account_components/basic_wallet.masl`: No such file or directory (os error 2)
[INFO] [stderr]  --> src/accounts/components/mod.rs:9:9
[INFO] [stderr]   |
[INFO] [stderr] 9 |         include_bytes!(concat!(env!("OUT_DIR"), "/assets/account_components/basic_wallet.masl"));
[INFO] [stderr]   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[INFO] [stderr]   |
[INFO] [stderr]   = note: this error originates in the macro `include_bytes` (in Nightly builds, run with -Z macro-backtrace for more info)
[INFO] [stderr] 
[INFO] [stderr] error: couldn't read `/opt/rustwide/target/x86_64-unknown-linux-gnu/debug/build/miden-lib-cfcaa074282a00e5/out/assets/account_components/rpo_falcon_512.masl`: No such file or directory (os error 2)
[INFO] [stderr]   --> src/accounts/components/mod.rs:16:9
[INFO] [stderr]    |
[INFO] [stderr] 16 |         include_bytes!(concat!(env!("OUT_DIR"), "/assets/account_components/rpo_falcon_512.masl"));
[INFO] [stderr]    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[INFO] [stderr]    |
[INFO] [stderr]    = note: this error originates in the macro `include_bytes` (in Nightly builds, run with -Z macro-backtrace for more info)
[INFO] [stderr] 
[INFO] [stderr] error: couldn't read `/opt/rustwide/target/x86_64-unknown-linux-gnu/debug/build/miden-lib-cfcaa074282a00e5/out/assets/account_components/basic_fungible_faucet.masl`: No such file or directory (os error 2)
[INFO] [stderr]   --> src/accounts/components/mod.rs:22:17
[INFO] [stderr]    |
[INFO] [stderr] 22 |       let bytes = include_bytes!(concat!(
[INFO] [stderr]    |  _________________^
[INFO] [stderr] 23 | |         env!("OUT_DIR"),
[INFO] [stderr] 24 | |         "/assets/account_components/basic_fungible_faucet.masl"
[INFO] [stderr] 25 | |     ));
[INFO] [stderr]    | |______^
[INFO] [stderr]    |
[INFO] [stderr]    = note: this error originates in the macro `include_bytes` (in Nightly builds, run with -Z macro-backtrace for more info)
[INFO] [stderr] 
[INFO] [stderr] error: couldn't read `/opt/rustwide/target/x86_64-unknown-linux-gnu/debug/build/miden-lib-cfcaa074282a00e5/out/assets/note_scripts/P2ID.masb`: No such file or directory (os error 2)
[INFO] [stderr]  --> src/notes/scripts.rs:9:17
[INFO] [stderr]   |
[INFO] [stderr] 9 |     let bytes = include_bytes!(concat!(env!("OUT_DIR"), "/assets/note_scripts/P2ID.masb"));
[INFO] [stderr]   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[INFO] [stderr]   |
[INFO] [stderr]   = note: this error originates in the macro `include_bytes` (in Nightly builds, run with -Z macro-backtrace for more info)
[INFO] [stderr] 
[INFO] [stderr] error: couldn't read `/opt/rustwide/target/x86_64-unknown-linux-gnu/debug/build/miden-lib-cfcaa074282a00e5/out/assets/note_scripts/P2IDR.masb`: No such file or directory (os error 2)
[INFO] [stderr]   --> src/notes/scripts.rs:16:17
[INFO] [stderr]    |
[INFO] [stderr] 16 |     let bytes = include_bytes!(concat!(env!("OUT_DIR"), "/assets/note_scripts/P2IDR.masb"));
[INFO] [stderr]    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[INFO] [stderr]    |
[INFO] [stderr]    = note: this error originates in the macro `include_bytes` (in Nightly builds, run with -Z macro-backtrace for more info)
[INFO] [stderr] 
[INFO] [stderr] error: couldn't read `/opt/rustwide/target/x86_64-unknown-linux-gnu/debug/build/miden-lib-cfcaa074282a00e5/out/assets/note_scripts/SWAP.masb`: No such file or directory (os error 2)
[INFO] [stderr]   --> src/notes/scripts.rs:23:17
[INFO] [stderr]    |
[INFO] [stderr] 23 |     let bytes = include_bytes!(concat!(env!("OUT_DIR"), "/assets/note_scripts/SWAP.masb"));
[INFO] [stderr]    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[INFO] [stderr]    |
[INFO] [stderr]    = note: this error originates in the macro `include_bytes` (in Nightly builds, run with -Z macro-backtrace for more info)
[INFO] [stderr] 
[INFO] [stderr] error: couldn't read `/opt/rustwide/target/x86_64-unknown-linux-gnu/debug/build/miden-lib-cfcaa074282a00e5/out/assets/note_scripts/P2ID.masb`: No such file or directory (os error 2)
[INFO] [stderr]   --> src/notes/utils.rs:18:17
[INFO] [stderr]    |
[INFO] [stderr] 18 |     let bytes = include_bytes!(concat!(env!("OUT_DIR"), "/assets/note_scripts/P2ID.masb"));
[INFO] [stderr]    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[INFO] [stderr]    |
[INFO] [stderr]    = note: this error originates in the macro `include_bytes` (in Nightly builds, run with -Z macro-backtrace for more info)
[INFO] [stderr] 
[INFO] [stderr] error: couldn't read `/opt/rustwide/target/x86_64-unknown-linux-gnu/debug/build/miden-lib-cfcaa074282a00e5/out/assets/kernels/tx_kernel.masl`: No such file or directory (os error 2)
[INFO] [stderr]   --> src/transaction/mod.rs:43:5
[INFO] [stderr]    |
[INFO] [stderr] 43 |     include_bytes!(concat!(env!("OUT_DIR"), "/assets/kernels/tx_kernel.masl"));
[INFO] [stderr]    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[INFO] [stderr]    |
[INFO] [stderr]    = note: this error originates in the macro `include_bytes` (in Nightly builds, run with -Z macro-backtrace for more info)
[INFO] [stderr] 
[INFO] [stderr] error: couldn't read `/opt/rustwide/target/x86_64-unknown-linux-gnu/debug/build/miden-lib-cfcaa074282a00e5/out/assets/kernels/tx_kernel.masb`: No such file or directory (os error 2)
[INFO] [stderr]   --> src/transaction/mod.rs:45:5
[INFO] [stderr]    |
[INFO] [stderr] 45 |     include_bytes!(concat!(env!("OUT_DIR"), "/assets/kernels/tx_kernel.masb"));
[INFO] [stderr]    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[INFO] [stderr]    |
[INFO] [stderr]    = note: this error originates in the macro `include_bytes` (in Nightly builds, run with -Z macro-backtrace for more info)
[INFO] [stderr] 
[INFO] [stderr] error: couldn't read `/opt/rustwide/target/x86_64-unknown-linux-gnu/debug/build/miden-lib-cfcaa074282a00e5/out/assets/miden.masl`: No such file or directory (os error 2)
[INFO] [stderr]   --> src/lib.rs:31:32
[INFO] [stderr]    |
[INFO] [stderr] 31 | const MIDEN_LIB_BYTES: &[u8] = include_bytes!(concat!(env!("OUT_DIR"), "/assets/miden.masl"));
[INFO] [stderr]    |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[INFO] [stderr]    |
[INFO] [stderr]    = note: this error originates in the macro `include_bytes` (in Nightly builds, run with -Z macro-backtrace for more info)
[INFO] [stderr] 
[INFO] [stderr] error: could not document `miden-lib`
[INFO] running `Command { std: "docker" "inspect" "9e2f96a3fb5156c28d1dd0a00b7c3d40424d15fe363836908b8b066285b45547", kill_on_drop: false }`
[INFO] running `Command { std: "docker" "rm" "-f" "9e2f96a3fb5156c28d1dd0a00b7c3d40424d15fe363836908b8b066285b45547", kill_on_drop: false }`
[INFO] [stdout] 9e2f96a3fb5156c28d1dd0a00b7c3d40424d15fe363836908b8b066285b45547
</code></pre>
</details>

So I think we need to fix our build process. We actually should not write into the `src` directory directly. I think `build.rs` files should always write into `OUT_DIR` (which can be written to in docs.rs environment) and then simply be included with something like `include!(concat!(env!("OUT_DIR"), "/build_file.rs"));`. docs.rs is setup with a read-only src directory because, I guess, it assumes that these best practices are followed. Us having problems with rustfmt and generated files in the past was basically also a symptom of us writing directly into `src`.
Off the top of my head this includes at least `kernel_v0.rs`, `tx_kernel_errors.rs` (which I created but did not follow the best practice since I also didn't know about this) and any files generated from protobuf files. Maybe others.

For reference, building with the docs.rs environment locally is a bit annoying but here are some pointers:
- Clone the docs.rs repo and setup the environment with this: https://github.com/rust-lang/docs.rs?tab=readme-ov-file#getting-started
- Before building a crate you need to package it (from the `miden-base` directory in this case), e.g.: `cargo package -p miden-lib --allow-dirty`. This is needed because that package is part of a workspace, see also: https://github.com/rust-lang/docs.rs/blob/master/docs/build-workspaces.md#when-do-we-need-this.
- And then build it from the docs.rs directory, e.g.: `cargo run -- build crate --local ../miden-base/target/package/miden-lib-0.6.1`.
- And then if you have the web-server started you can find the miden-lib build and see if it failed there.

Hope this helps!

igamigo commented:
I see. We had tried to build the image and run everything but it was taking too long, so we went with a familiar approach (this problem came up some months ago). I'll give it a second try and test your approach (wasn't keeping in mind that we should be able to write to `OUT_DIR`). Thanks!

igamigo commented:
Hmm, I spent some time trying to run the docs.rs pipeline (with the Docker alternative that they provide for MacOS instead of running it natively), but was not able to. I was able to sort through a couple minor issues but the image does not build correctly. Specifically, it fails to build the dummy crate with "failed to add essential files".

I decided to not waste too much more time on that but test your suggested approach anyway. 
I think this works in general, although unless I misunderstood something, these files specifically (`kernel_v0.rs` for example) did not seem to be the reason why the docs were failing to build according to the logs you pasted. It seems the main problem there was that with these changes, the kernel and note scripts were not being built (because the docs.rs check was being done before those) and so doing `include_bytes!(concat!(env!("OUT_DIR"), "P2ID.masb"))` would fail.

The main consequence of your approach is that code becomes a bit harder to browse (and less flexible in general) due to the `include!()`s, and files not being committed. So I guess for any build script, this could be an alternative:

1 - Build any binary file to `env!("OUT_DIR")` that we are including with `include_bytes!()`
2 - Check whether we are on the docs.rs pipeline and return if so
3 - Build any `.rs` file that we want to place under `/src` then. These files get committed to the source tree anyway

PhilippGackstatter commented on Makefile:
```diff
@@ -15,7 +15,7 @@ BUILD_KERNEL_ERRORS=BUILD_KERNEL_ERRORS=1
 .PHONY: clippy
 clippy: ## Runs Clippy with configs
-	cargo clippy --workspace --all-targets $(ALL_FEATURES_BUT_ASYNC) -- -D warnings
+	${BUILD_KERNEL_ERRORS} cargo clippy --workspace --all-targets $(ALL_FEATURES_BUT_ASYNC) -- -D warnings
```
This env var seems no longer used now.

igamigo replied:
Resolving this as the approach changed.

PhilippGackstatter commented on miden-lib/build.rs:
```diff
@@ -37,6 +37,12 @@ fn main() -> Result<()> {
     println!("cargo:rerun-if-changed=asm");
     println!("cargo:rerun-if-changed={KERNEL_V0_RS_FILE}");
 
+    // The docs.rs build pipeline has a read-only filesystem, so we want to return early or
+    // otherwise the docs will fail to build there.
+    if env::var("DOCS_RS").is_ok() {
+        return Ok(());
+    }
+
```
If we remove this, then the docs build succeeds (tested it locally).

This is fine since this build script now no longer writes to `src` but only `OUT_DIR`.

igamigo replied:
Resolving this as the approach changed.

PhilippGackstatter commented on miden-tx/build.rs:
```diff
@@ -18,13 +17,6 @@ const MASM_FILE_EXTENSION: &str = "masm";
 fn main() -> Result<()> {
     // re-build when the MASM code changes
     println!("cargo:rerun-if-changed={ASM_DIR}");
-    println!("cargo:rerun-if-changed={KERNEL_ERRORS_FILE}");
-    println!("cargo::rerun-if-env-changed=BUILD_KERNEL_ERRORS");
-
-    // Skip this build script in BUILD_KERNEL_ERRORS environment variable is not set to `1`.
-    if env::var("BUILD_KERNEL_ERRORS").unwrap_or("0".to_string()) == "0" {
-        return Ok(());
-    }
```
Can you also update the doc comment of the main function here to remove the reference to this env var?

igamigo replied:
Marking this as resolved since the approach changed.

PhilippGackstatter reviewed (approved):
Looks good to me.

It seems like best practice to write to `OUT_DIR` instead of `src` so I think those changes are good. In my IDE (Vs Code) I can still jump to the built file if I click on a symbol defined there, so it seems okay, but it's not ideal. And you're right, the build could actually work with the checked-in files as well - I didn't think of this somehow.

With the env-check in miden-lib removed the docs.rs build works.

I also tested building miden-tx-prover, and it seems to work in terms of not writing to `src` but it fails because `cmake` is missing. I don't know if this is a quirk of the local setup because cmake should be such a common dependency that I assume it would be available on docs.rs itself (e.g. https://github.com/rust-lang/crates-build-env/issues/160#issuecomment-2415807441). So perhaps this is fine and we can just try another patch release to see if that works.

<details>
<summary>miden-tx-provder build output logs</summary>
<code><pre>
# rustc version
rustc 1.84.0-nightly (5ec7d6eee 2024-11-17)# docs.rs version
docsrs 0.6.0 (e2e7e30 2024-11-18)# build log
[INFO] running `Command { std: "docker" "create" "-v" "~/git/docs.rs/.workspace/builds/miden-tx-prover-0.6.1/target:/opt/rustwide/target:rw,Z" "-v" "~/git/docs.rs/.workspace/builds/miden-tx-prover-0.6.1/source:/opt/rustwide/workdir:ro,Z" "-v" "~/git/docs.rs/.workspace/cargo-home:/opt/rustwide/cargo-home:ro,Z" "-v" "~/git/docs.rs/.workspace/rustup-home:/opt/rustwide/rustup-home:ro,Z" "-e" "SOURCE_DIR=/opt/rustwide/workdir" "-e" "CARGO_TARGET_DIR=/opt/rustwide/target" "-e" "DOCS_RS=1" "-e" "CARGO_HOME=/opt/rustwide/cargo-home" "-e" "RUSTUP_HOME=/opt/rustwide/rustup-home" "-w" "/opt/rustwide/workdir" "-m" "3221225472" "--user" "1000:1000" "--network" "none" "ghcr.io/rust-lang/crates-build-env/linux-micro" "/opt/rustwide/cargo-home/bin/cargo" "+nightly" "rustdoc" "--lib" "-Zrustdoc-map" "--config" "build.rustdocflags=[\"--cfg\", \"docsrs\", \"-Z\", \"unstable-options\", \"--emit=invocation-specific\", \"--resource-suffix\", \"-20241117-1.84.0-nightly-5ec7d6eee\", \"--static-root-path\", \"/-/rustdoc.static/\", \"--cap-lints\", \"warn\", \"--extern-html-root-takes-precedence\"]" "--offline" "-Zunstable-options" "--config=doc.extern-map.registries.crates-io=\"https://docs.rs/{pkg_name}/{version}/x86_64-unknown-linux-gnu\"" "-Zrustdoc-scrape-examples" "--target" "x86_64-unknown-linux-gnu", kill_on_drop: false }`
[INFO] [stdout] 67006250182d8afee8835735dfb0f91a3b57b04eb221c7c50f95e6ea7e7e578c
[INFO] running `Command { std: "docker" "start" "-a" "67006250182d8afee8835735dfb0f91a3b57b04eb221c7c50f95e6ea7e7e578c", kill_on_drop: false }`
[INFO] [stderr] warning: target filter specified, but no targets matched; this is a no-op
[INFO] [stderr]    Compiling miden-assembly v0.11.0
[INFO] [stderr]    Compiling zstd-safe v7.2.1
[INFO] [stderr]    Compiling openssl v0.10.68
[INFO] [stderr]     Checking hyper v0.14.31
[INFO] [stderr]     Checking zstd-sys v2.0.13+zstd.1.5.6
[INFO] [stderr]     Checking openssl-sys v0.9.104
[INFO] [stderr]    Compiling tokio-openssl v0.6.5
[INFO] [stderr]    Compiling libz-ng-sys v1.1.20
[INFO] [stderr]     Checking pingora-error v0.3.0
[INFO] [stderr]     Checking ipnet v2.10.1
[INFO] [stderr]     Checking tower-layer v0.3.3
[INFO] [stderr] error: failed to run custom build command for `libz-ng-sys v1.1.20`
[INFO] [stderr] 
[INFO] [stderr] Caused by:
[INFO] [stderr]   process didn't exit successfully: `/opt/rustwide/target/debug/build/libz-ng-sys-45397d3369b770c2/build-script-cmake` (exit status: 101)
[INFO] [stderr]   --- stdout
[INFO] [stderr]   CMAKE_TOOLCHAIN_FILE_x86_64-unknown-linux-gnu = None
[INFO] [stderr]   CMAKE_TOOLCHAIN_FILE_x86_64_unknown_linux_gnu = None
[INFO] [stderr]   HOST_CMAKE_TOOLCHAIN_FILE = None
[INFO] [stderr]   CMAKE_TOOLCHAIN_FILE = None
[INFO] [stderr]   CMAKE_GENERATOR_x86_64-unknown-linux-gnu = None
[INFO] [stderr]   CMAKE_GENERATOR_x86_64_unknown_linux_gnu = None
[INFO] [stderr]   HOST_CMAKE_GENERATOR = None
[INFO] [stderr]   CMAKE_GENERATOR = None
[INFO] [stderr]   CMAKE_PREFIX_PATH_x86_64-unknown-linux-gnu = None
[INFO] [stderr]   CMAKE_PREFIX_PATH_x86_64_unknown_linux_gnu = None
[INFO] [stderr]   HOST_CMAKE_PREFIX_PATH = None
[INFO] [stderr]   CMAKE_PREFIX_PATH = None
[INFO] [stderr]   CMAKE_x86_64-unknown-linux-gnu = None
[INFO] [stderr]   CMAKE_x86_64_unknown_linux_gnu = None
[INFO] [stderr]   HOST_CMAKE = None
[INFO] [stderr]   CMAKE = None
[INFO] [stderr]   running: cd "/opt/rustwide/target/x86_64-unknown-linux-gnu/debug/build/libz-ng-sys-2cbf93fb004c50b8/out/build" && CMAKE_PREFIX_PATH="" "cmake" "/opt/rustwide/cargo-home/registry/src/index.crates.io-6f17d22bba15001f/libz-ng-sys-1.1.20/src/zlib-ng" "-DBUILD_SHARED_LIBS=OFF" "-DZLIB_COMPAT=OFF" "-DZLIB_ENABLE_TESTS=OFF" "-DWITH_GZFILEOP=ON" "-DCMAKE_INSTALL_PREFIX=/opt/rustwide/target/x86_64-unknown-linux-gnu/debug/build/libz-ng-sys-2cbf93fb004c50b8/out" "-DCMAKE_C_FLAGS= -ffunction-sections -fdata-sections -fPIC -m64" "-DCMAKE_C_COMPILER=/usr/bin/cc" "-DCMAKE_CXX_FLAGS= -ffunction-sections -fdata-sections -fPIC -m64" "-DCMAKE_CXX_COMPILER=/usr/bin/c++" "-DCMAKE_ASM_FLAGS= -ffunction-sections -fdata-sections -fPIC -m64" "-DCMAKE_ASM_COMPILER=/usr/bin/cc" "-DCMAKE_BUILD_TYPE=Debug"
[INFO] [stderr] 
[INFO] [stderr]   --- stderr
[INFO] [stderr]   thread 'main' panicked at /opt/rustwide/cargo-home/registry/src/index.crates.io-6f17d22bba15001f/cmake-0.1.51/src/lib.rs:1100:5:
[INFO] [stderr] 
[INFO] [stderr]   failed to execute command: No such file or directory (os error 2)
[INFO] [stderr]   is `cmake` not installed?
[INFO] [stderr] 
[INFO] [stderr]   build script failed, must exit now
[INFO] [stderr]   note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[INFO] [stderr] warning: build failed, waiting for other jobs to finish...
[INFO] running `Command { std: "docker" "inspect" "67006250182d8afee8835735dfb0f91a3b57b04eb221c7c50f95e6ea7e7e578c", kill_on_drop: false }`
[INFO] running `Command { std: "docker" "rm" "-f" "67006250182d8afee8835735dfb0f91a3b57b04eb221c7c50f95e6ea7e7e578c", kill_on_drop: false }`
[INFO] [stdout] 67006250182d8afee8835735dfb0f91a3b57b04eb221c7c50f95e6ea7e7e578c
</pre></code>
</details>

For miden-tx we have a problem because the build script uses a relative path to miden-lib which is not available after packaging the crate. So the build fails with a  `| miden-tx-0.6.1/../miden-lib/asm: No such file or directory (os error 2) -> No such file or directory (os error 2)`. When I introduced the error build I wanted to avoid building the errors in miden-lib because we only need them in miden-tx so I used the relative path without realizing it would cause these problems. Now I think, given the constraints here, we should just build the tx kernel errors in miden-lib and export them from there.

I'll approve this PR since we seem to have fixed the writing to src issue, but we will need at least one follow-up PR. I'm happy to take that issue on the tx kernel change since I can test it locally. So I'll approve in case you want to merge the PR now. Let me know how you want to proceed and whether I should create an issue for this.

igamigo commented:
I changed the approach back to what it was earlier, but only avoiding to build the Rust files specifically. I also moved the kernel error generation to `miden-lib` as suggested by Bobbin, and tweaked the CI check.

bobbinth commented on miden-lib/Cargo.toml:
```diff
@@ -34,3 +34,4 @@ vm-processor = { workspace = true, features = ["testing"] }
 assembly = { workspace = true }
 miden-stdlib = { workspace = true }
 regex = { version = "1.10" }
+walkdir = { version = "2.5"}
```
nit: let's add a space between "2.5" and `}`.

bobbinth commented on miden-lib/src/errors/mod.rs:
```diff
@@ -0,0 +1,3 @@
+#[rustfmt::skip]
+mod tx_kernel_errors;
+pub use tx_kernel_errors::*;
```
Would it make sense to export the module itself? Basically like:
```rust
#[rustfmt::skip]
pub mod tx_kernel_errors;
```
This would make error constants more explicit, which I think is a good thing.

igamigo replied:
I did this originally but the fully qualified path for an error code seemed a bit long and a bit redundant as well, but I can change it back

igamigo replied:
Changed this

bobbinth reviewed (approved):
Looks good! Thank you! I left a couple small comments inline.

bobbinth commented:
One other thing: maybe we should merge this into `main`? This way, I could make a patch release and this should allow docs to be built on `docs.rs`.

igamigo commented:
@bobbinth if you agree with testing with a patch release we can go ahead and merge this

bobbinth commented:
> @bobbinth if you agree with testing whether the error that Philipp reported does not happen with a patch release we can go ahead and merge this

Yes, let's merge. I'll be able to do a patch release later today.

bobbinth commented:
I published v0.6.2 crates and it seems like `miden-lib` [docs](https://docs.rs/crate/miden-lib/0.6.2) are now building fine.
